README.rst
===========
rio-ds-mask
===========

A `Rasterio <https://github.com/mapbox/rasterio>`__ plugin for extracting
an image's dataset-level mask.

.. image:: https://travis-ci.org/geowurster/rio-ds-mask.svg?branch=master
    :target: https://travis-ci.org/geowurster/rio-ds-mask

.. image:: https://coveralls.io/repos/github/geowurster/rio-ds-mask/badge.svg?branch=master
    :target: https://coveralls.io/github/geowurster/rio-ds-mask


=====
Usage
=====

.. code-block:: console

    Usage: rio ds-mask [OPTIONS] INPUT OUTPUT

      Extract an image's dataset-level mask.

      Both output driver and datatype are derived from the input image if not
      given.

      In some cases this plugin alters GDAL's returned mask values.  When
      writing masks GDAL uses 0's for opaque and 255's for transparent, but when
      reading masks the returned value differs based on the image's datatype.  8
      bit images produce 8 bit masks where 0's are opaque and 255's are
      transparent, however 16 bit images use 0's for opaque and 1's for
      transparent, still stored as 8 bit.  If the image's datatype is 'int16' or
      'uint16' and the mask's maximum value is 1, then all 1's are translated to
      255's.  The mask's datatype is preserved.  I have not fully investigated
      all of GDAL's masking options to determine if the behavior is consistent.
      If it is found to be a deliberate choice then the normalization will be
      removed.

    Options:
      -f, --format, --driver TEXT     Output format driver
      -t, --dtype [ubyte|uint8|uint16|int16|uint32|int32|float32|float64]
                                      Output data type.
      --co, --profile NAME=VALUE      Driver specific creation options.See the
                                      documentation for the selected output driver
                                      for more information.
      --help                          Show this message and exit.

This example command creates a singleband ``uint8`` image that is acceptable
to use as a GDAL mask band, meaning that pixels with a value of ``255`` are
transparent and pixels with a vaue of ``0`` are opaque.  The image is
losslessly compressed and internally tiled.

.. code-block:: console

    $ rio ds-mask \
        --driver GTiff \
        tests/data/alpha.tif \
        mask.tif \
        --co COMPRESS=DEFLATE \
        --co TILED=YES


==========
Installing
==========

First `install Rasterio <http://mapbox.github.io/rasterio/installation.html>`__,
then:

.. code-block:: console

    $ pip install rio-ds-mask --user


Developing
==========

.. code-block:: console

    $ git clone https://github.com/geowurster/rio-ds-mask.git
    $ cd rio-ds-mask
    $ pip install -e .\[all\]
    $ pytest --cov rio-ds-mask --cov-report term-missing


License
=======

See `LICENSE.txt <LICENSE.txt>`__


Changelog
=========

See `CHANGES.md <CHANGES.md>`__


CHANGES.md
Changelog
=========

1.2 (2018-02-10)
----------------

- Detect output datatype from mask if `--dtype` is not given (#3).
- Normalize GDAL masks using 1's as transparent to 255's (#4).


1.1 (2017-12-11)
----------------

- Small packaging fixes.


1.0 (2017-12-11)
----------------

- Initial release.


tests/conftest.py
"""``pytest`` fixtures."""


import os
import shutil

from click.testing import CliRunner
import pytest


@pytest.fixture(scope='function')
def path_alpha_tif(tmpdir):

    """Path to 4 band test file with an alpha band."""

    outfile = str(tmpdir.join('alpha.tif'))
    shutil.copy(os.path.join('tests', 'data', 'alpha.tif'), outfile)
    return outfile


@pytest.fixture(scope='function')
def path_alpha_16bit_tif(tmpdir):

    """Path to 4 band 16 bit test file with an alpha band."""

    outfile = str(tmpdir.join('alpha_16bit.tif'))
    shutil.copy(os.path.join('tests', 'data', 'alpha_16bit.tif'), outfile)
    return outfile


@pytest.fixture(scope='function')
def runner():

    """A standard ``click.testing.CliRunner()``."""

    return CliRunner()


tests/test_cli.py
"""Tests for ``$ rio ds-mask`` plugin."""


import numpy as np
import rasterio as rio
from rasterio.enums import Compression
from rasterio.rio.main import main_group as rio_main

from rio_ds_mask import _norm_gdal_mask, rio_ds_mask


def test_compare_rasterio_dataset_mask(
        path_alpha_tif, path_alpha_16bit_tif, runner, tmpdir):

    """Compare against Rasterio's ``src.dataset_mask()``."""

    outfile = str(tmpdir.join('compare_rasterio_dataset_mask.tif'))

    for path in (path_alpha_tif, path_alpha_16bit_tif):
        with rio.open(path) as src:
            expected = _norm_gdal_mask(src.dataset_mask(), src.dtypes)

        result = runner.invoke(rio_ds_mask, [
            path, outfile,
            '--co', 'COMPRESS=DEFLATE'])

        assert result.exit_code == 0

        with rio.open(outfile) as src:
            assert src.compression == Compression.deflate
            assert not src.is_tiled
            actual = src.read(1)

        assert np.array_equal(expected, actual)


def test_dtype(runner, tmpdir, path_alpha_16bit_tif):

    """Set output datatype and driver."""

    outfile = str(tmpdir.join('test_dtype_driver.jpg'))

    result = runner.invoke(rio_ds_mask, [
        path_alpha_16bit_tif, outfile,
        '--dtype', 'uint8',
        '--driver', 'JPEG'])

    assert result.exit_code == 0, result.output

    with rio.open(outfile) as src:
        assert src.dtypes[0] == rio.uint8
        assert src.driver == 'JPEG'


def test_16bit_uses_255(runner, tmpdir, path_alpha_16bit_tif):

    """For 16 bit images GDAL produces masks with transparent values as 1's
    rather than 255's.  Verify this plugin is translating those values to
    255's.
    """

    outfile = str(tmpdir.join('test_16bit_uses_255.tif'))

    result = runner.invoke(rio_ds_mask, [
        path_alpha_16bit_tif, outfile])
    assert result.exit_code == 0

    with rio.open(outfile) as src:
        mask = src.read(1)
        assert mask.dtype == np.uint8
        assert mask.min() == 0
        assert mask.max() == 255


def test_plugin_registered(runner):

    """Make sure plugin is registered."""

    result = runner.invoke(rio_main, ['--help'])
    assert result.exit_code == 0
    assert 'ds-mask' in result.output
    assert rio_ds_mask.short_help in result.output


tests/test_core.py
"""Tests for non-cli features."""


import pytest
import numpy as np
from rio_ds_mask import _norm_gdal_mask


@pytest.mark.parametrize("mask,expected", [
    # Totally masked
    (np.array([0, 0]), np.array([0, 0])),
    # Partially masked + trigger heuristic
    (np.array([0, 1]), np.array([0, 255])),
    # Partially masked with 255's instead of 1's shouldn't be altered
    (np.array([0, 255]), np.array([0, 255])),
    # Totally transparent
    (np.array([1, 1]), np.array([255, 255])),
    # Varying alpha values shouldn't be altered
    (np.array([0, 1, 2]), np.array([0, 1, 2])),
    # Values past 255 shouldn't be altered
    (np.array([0, 256]), np.array([0, 256])),
])
def test_norm_gdal_mask(mask, expected):

    """Test for ``rio_ds_mask._norm_gdal_mask()``.  Tests both ``np.uint16``
    and ``np.int16``.
    """

    for dtype in (np.uint16, np.int16):
        mask = _norm_gdal_mask(
            np.array(mask, dtype=dtype),
            [np.dtype(dtype).name])
        assert np.array_equal(mask, expected)
        assert mask.dtype == dtype


rio_ds_mask.py
"""Extract an image's dataset-level mask and write to a new file."""


__version__ = '1.2'
__author__ = 'Kevin Wurster'
__email__ = 'wursterk@gmail.com'
__source__ = 'https://github.com/geowurster/rio-ds-mask'
__license__ = '''
New BSD License

Copyright (c) 2017, Kevin D. Wurster
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* The names of rio-ds-mask or its contributors may not be used to endorse or
  promote products derived from this software without specific prior written
  permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'''


import itertools as it
import operator as op
import sys

import click
import cligj
import numpy as np
import rasterio as rio
from rasterio.rio import options


if sys.version_info[0] == 2:
    from itertools import imap as map


def _norm_gdal_mask(array, src_dtypes):

    """When writing masks GDAL uses 0's for opaque and 255's for transparent,
    but when reading masks the returned value differs based on the image's
    datatype.  8 bit images produce masks where 0's are opaque and 255's
    are transparent, however 16 bit images use 0's for opaque and 1's for
    transparent.  If the image's datatype is 'int16' or 'uint16' and the
    mask's maximum value is 1, then all 1's are translated to 255's.  The
    mask's datatype is preserved.  I have not fully investigated all of
    GDAL's masking options to determine if the behavior is consistent.  If
    it is found to be a deliberate choice then the normalization will be
    removed.

    Parameters
    ----------
    array : numpy.ndarray
        Array to normalize.
    src_dtypes : sequence
        Parent image's datatypes.

    Returns
    -------
    numpy.ndarray
    """

    cast = getattr(np, array.dtype.name)
    if set(src_dtypes) & {'uint16', 'int16'} and array.max() == 1:
        array = np.where(array > 0, cast(255), cast(0))
    return array


@click.command(name='ds-mask')
@options.file_in_arg
@options.file_out_arg
@cligj.format_opt
@options.dtype_opt
@options.creation_options
def rio_ds_mask(input, output, driver, dtype, creation_options):

    """Extract an image's dataset-level mask.

    Both output driver and datatype are derived from the input image if
    not given.

    In some cases this plugin alters GDAL's returned mask values.  When
    writing masks GDAL uses 0's for opaque and 255's for transparent, but
    when reading masks the returned value differs based on the image's
    datatype.  8 bit images produce 8 bit masks where 0's are opaque and
    255's are transparent, however 16 bit images use 0's for opaque and 1's
    for transparent, still stored as 8 bit.  If the image's datatype is
    'int16' or 'uint16' and the mask's maximum value is 1, then all 1's are
    translated to 255's.  The mask's datatype is preserved.  I have not fully
    investigated all of GDAL's masking options to determine if the behavior
    is consistent.  If it is found to be a deliberate choice then the
    normalization will be removed.
    """

    with rio.open(input) as src:

        all_windows = map(op.itemgetter(1), src.block_windows())
        window_data = ((w, src.dataset_mask(window=w)) for w in all_windows)

        first = next(window_data)
        window_data = it.chain([first], window_data)
        detected_dtype = first[1].dtype.name
        del first

        meta = src.meta.copy()
        meta.update(
            count=1,
            driver=driver or src.driver,
            dtype=dtype or detected_dtype)

        if creation_options:
            meta.update(**creation_options)

        with rio.open(output, 'w', **meta) as dst:
            for window, mask in window_data:
                dst.write(
                    _norm_gdal_mask(mask, src.dtypes), indexes=1, window=window)
